apollo-router/tests/integration/subgraph_response.rs

```diff
@@ -9,6 +9,78 @@ include_subgraph_errors:
   all: true
 "#;
 
+#[tokio::test(flavor = "multi_thread")]
+async fn test_subgraph_returning_data_null() -> Result<(), BoxError> {
+    let mut router = IntegrationTest::builder()
+        .config(CONFIG)
+        .responder(ResponseTemplate::new(200).set_body_json(json!({ "data": null })))
+        .build()
+        .await;
+
+    router.start().await;
+    router.assert_started().await;
+
+    let query = "{ __typename topProducts { name } }";
+    let (_trace_id, response) = router.execute_query(&json!({ "query":  query })).await;
+    assert_eq!(response.status(), 200);
+    assert_eq!(
+        serde_json::from_str::<serde_json::Value>(&response.text().await?)?,
+        json!({ "data": null })
```

Comment:
why should `{ __typename topProducts { name } }` return a null `data` if the subgraph returns null? `topProducts` is a nullable field, right?
Is it for the case where the subgraph would nullify because there was a non nullable field?

Reply:
You are right, it was mainly to test null propagation that goes up to `data.`
However, graphql servers can return `data: null` if execution is started by no resolvers executed:
https://spec.graphql.org/October2021/#sel-FAPHPHCAACEBr6C

For example, here is the test from graphql-js:
https://github.com/graphql/graphql-js/blob/993d7ce2ee6db2a13973b037817f3eac60607b8a/src/execution/__tests__/executor-test.ts#L971-L981

These cases are but could happen, so we should keep semantics.

Reply:
I guess the "right" way would be to detect from errors if we got a null because a non null field was nullified by the subgraph, but that may be too complex